### PR TITLE
refactor: create a storage crate for snapshot

### PR DESCRIPTION
**Pull request by liubaohai**
# Description           

- snapshot提取到一个独立的crate storage
- storage这个crate不依赖其他crate，因此部分和epaxos有关的接口留在了epaxos里面
- 完善了部分测试用例

## Type of change       

- **Refactoring**

# Checklist:

- [ ] **Self-review**: I have performed a **self-review** of my own code
- [ ] **Comment**:     I have **commented my code**, particularly in hard-to-understand areas
- [ ] **Doc**:         I have made corresponding changes to the **documentation**
- [ ] **No-warnings**: My changes generate **no new warnings**
- [x] **Add-test**:    I have added **tests** that prove my fix is effective or that my feature works
- [x] **Pass**:        New and existing **unit tests pass** locally with my changes

### CODE REVIEW

File: components/epaxos/src/lib.rs [hunk 1]
```diff
@@ -14,8 +14,47 @@ mod service;
...
+impl From<Instance> for WriteEntry {
+    fn from(inst: Instance) -> Self {
+        let mut v = vec![];
+        inst.encode(&mut v).unwrap();
+        return WriteEntry::Set(DBColumnFamily::Instance, inst.to_key(), v);
+    }
+}
+
+// just for set max exec ref
+impl From<InstanceId> for WriteEntry {
```
Comment (liubaohai) [inline components/epaxos/src/lib.rs:52]: 很无奈这里只能这样写，因为WriteEntry，（&str, Into<>）这两个类型都不属于本crate。

Comment (drmingdrmer) [reply components/epaxos/src/lib.rs:52]: > 很无奈这里只能这样写，因为WriteEntry，（&str, Into<>）这两个类型都不属于本crate。


你这句话放代码注释里吧...

File: components/epaxos/src/replication/errors.rs [hunk 2]
```diff
@@ -3,7 +3,7 @@ use crate::qpaxos::ProtocolError;
 use crate::qpaxos::QError;
 use crate::qpaxos::ReplicaID;
 use crate::replica::Error as ReplicaError;
-use crate::snapshot::Error as SnapError;
+use storage::Error as StorageError;
```
Comment (drmingdrmer) [inline components/epaxos/src/replication/errors.rs:6]: storage::Error 直接定义长storage::StorageError是否好点?  我也不太确定, as重命名可能在不同的文件里出现不同的名字, 之后写代码可能有有1丢丢麻烦

File: components/storage/src/test_engine.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
+use crate::*;
+use prost::Message;
+
+#[derive(Clone, PartialEq, Message, Copy, Eq, Ord, PartialOrd, Hash)]
+pub struct TestInstance {
+    #[prost(int64, tag = "1")]
+    pub id: i64,
+    #[prost(int64, tag = "2")]
+    pub foo: i64,
+}
```
Comment (drmingdrmer) [inline components/storage/src/test_engine.rs:10]: `∑ ˙∆˚`...还有这种操作

File: components/epaxos/src/iters.rs [hunk 4]
```diff
@@ -1,25 +1,24 @@
-use super::traits::*;
 use crate::qpaxos::*;
-use crate::tokey::ToKey;
+use crate::Storage;
```
[the remark this answers is not in the document]
Comment (liubaohai) [reply components/epaxos/src/iters.rs:2]: 不是同一个，这个类型定义在当前crate

**Pull request merged**